gprotorch/kernels/gnn_kernels/test_hu_iclr_2020.py

```diff
@@ -0,0 +1,357 @@
+        Chem.rdchem.BondType.TRIPLE,
+        Chem.rdchem.BondType.AROMATIC,
+    ],
+    # (E)/(Z) double bond stereo information
+    "possible_bond_dirs": [
+        Chem.rdchem.BondDir.NONE,
+        Chem.rdchem.BondDir.ENDUPRIGHT,
+        Chem.rdchem.BondDir.ENDDOWNRIGHT,
+    ],
+}
+
+# define additional tokens for label masking
+num_atom_type = 120  # including the extra mask tokens
+num_chirality_tag = 3
+
+num_bond_type = (
+    6  # including aromatic and self-loop edge, and extra masked tokens
+)
+num_bond_direction = 3
+self_loop_token = 4  # bond type for self-loop edge
+masked_bond_token = 5  # bond type for masked edges
+
+
+def mol_to_pyg(mol):
+    """
+    A featuriser that accepts an rdkit mol instance and
+    converts it to a PyTorch Geometric data object that
+    is compatible with the GNN modules below.
+
+    Args:
+        mol: rdkit mol object
+
+    Returns: PyTorch Geometric data object
+
+    """
+
+    # derive atom features: atomic number + chirality tag
+    atom_features = []
+    for atom in mol.GetAtoms():
+        atom_features.append(
+            [
+                allowable_features["possible_atomic_num_list"].index(
+                    atom.GetAtomicNum()
+                ),
+                allowable_features["possible_chirality_list"].index(
+                    atom.GetChiralTag()
+                ),
+            ]
+        )
+    atom_features = torch.tensor(np.array(atom_features), dtype=torch.long)
+
+    # derive bond features: bond type + bond direction
+    # PyTorch Geometric only uses directed edges,
+    # so feature information needs to be added twice
+    edge_index = []
+    edge_attr = []
+    for bond in mol.GetBonds():
+        i = bond.GetBeginAtomIdx()
+        j = bond.GetEndAtomIdx()
+        edge_index.append((i, j))
+        edge_index.append((j, i))
+
+        # calculate edge features and append them to feature list
+        edge_feature = [
+            allowable_features["possible_bonds"].index(bond.GetBondType()),
+            allowable_features["possible_bond_dirs"].index(bond.GetBondDir()),
+        ]
+        edge_attr.append(edge_feature)
+        edge_attr.append(edge_feature)
+
+    # set data.edge_index: Graph connectivity in COO format with shape [2, num_edges]
+    edge_index = torch.tensor(np.array(edge_index).T, dtype=torch.long)
+
+    # set data.edge_attr: Edge feature matrix with shape [num_edges, num_edge_features]
+    edge_attr = torch.tensor(np.array(edge_attr), dtype=torch.long)
+
+    return Data(x=atom_features, edge_index=edge_index, edge_attr=edge_attr)
+
+
+class GINConv(MessagePassing):
+    """
+    Extension of the Graph Isomorphism Network to incorporate
+    edge information by concatenating edge embeddings.
+    """
+
+    def __init__(self, emb_dim, aggr="add"):
+        """
+        Initialise GIN convolutional layer.
+        Args:
+            emb_dim: latent node embedding dimension
+            aggr: aggregation procedure
```

Comment:
What are the different options for the aggregation procedure out of interest?

Reply:
For GINs you can also do mean pooling and element-wise max pooling.